include/kernels/dslash_ndeg_twisted_clover.cuh

```diff
@@ -0,0 +1,112 @@
+      // if dagger flip the chiral twist
+      // factor of 1/2 comes from clover normalization 
+      b(dagger ? -0.5 * b : 0.5 * b),
+      c(c)
+      {
+        checkPrecision(U, A);
+        checkLocation(U, A);
+      }
+  };
+  
+  template <int nParity, bool dagger, bool xpay, KernelType kernel_type, typename Arg>
+    struct nDegTwistedClover : dslash_default {
+    
+    Arg &arg;
+    constexpr nDegTwistedClover(Arg &arg) : arg(arg) {}
+    static constexpr const char *filename() { return KERNEL_FILE; } // this file name - used for run-time compilation
+    
+    /**
+       @brief Apply the non-degenerate twisted-clover dslash
+       out(x) = M*in = A(x)*in(x) + a * D * in + (1 + i*b*gamma_5*tau_3 + c*tau_1)*in
+       Note this routine only exists in xpay form.
+    */
+    __device__ __host__ inline void operator()(int idx, int flavor, int parity)
+    {
+      typedef typename mapper<typename Arg::Float>::type real;
+      typedef ColorSpinor<real, Arg::nColor, 4> Vector;
+      typedef ColorSpinor<real, Arg::nColor, 2> HalfVector;
+      
+      bool active
+        = kernel_type == EXTERIOR_KERNEL_ALL ? false : true; // is thread active (non-trival for fused kernel only)
+      int thread_dim;                                        // which dimension is thread working on (fused kernel only)
+
+      // FIXME: do we actually need two of these?
+      // checking getCoords, the 's' parameter (where we pass flavor) simply sets coord.s
+      // such that we likely don't need to do this, as the 4D coords are unchanged
+      auto clover_coord = getCoords<QUDA_4D_PC, kernel_type>(arg, idx, 0, parity, thread_dim);
+      auto coord = getCoords<QUDA_4D_PC, kernel_type>(arg, idx, flavor, parity, thread_dim);
+      
+      const int my_spinor_parity = nParity == 2 ? parity : 0;
+      Vector out;
+      
+      // defined in dslash_wilson.cuh
+      applyWilson<nParity, dagger, kernel_type>(out, arg, coord, parity, idx, thread_dim, active);
+
+      int my_flavor_idx = coord.x_cb + flavor * arg.dc.volume_4d_cb;
+
+      if (kernel_type == INTERIOR_KERNEL) {
+        // apply the chiral and flavor twists
+        // use consistent load order across s to ensure better cache locality
+        Vector x0 = arg.x(coord.x_cb + 0 * arg.dc.volume_4d_cb, my_spinor_parity);
+        Vector x1 = arg.x(coord.x_cb + 1 * arg.dc.volume_4d_cb, my_spinor_parity);
+
+        Vector x = flavor == 0 ? x0 : x1;
+        x.toRel(); // switch to chiral basis
+        
+        Vector tmp;
+#pragma unroll
+        for (int chirality = 0; chirality < 2; chirality++) {
+          constexpr int n = Arg::nColor * Arg::nSpin / 2;
+          HMatrix<real, n> A = arg.A(clover_coord.x_cb, parity, chirality);
+          HalfVector x_chi = x.chiral_project(chirality);
+          HalfVector Ax_chi = A * x_chi;
+          // i * mu * gamma_5 * tau_3
+          const complex<real> b(0.0, (chirality^flavor) == 0 ? static_cast<real>(arg.b) : -static_cast<real>(arg.b));
```

Comment:
@kostrzewa do you agree on this line? Before the sign change due to flavor was missing.

Reply:
yes, 0^0 = 0, 1^0 = 0^1 = 1, 1^1 = 0

Reply:
and as far as I can tell, the sign of mu is not adjusted further up the call graph